pkg/ebpf/perf/event.go

```diff
@@ -0,0 +1,228 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024-present Datadog, Inc.
+
+//go:build linux_bpf
+
+// Package perf implements types related to eBPF and the perf subsystem, like perf buffers and ring buffers.
+package perf
+
+import (
+	"errors"
+	"fmt"
+	"slices"
+
+	manager "github.com/DataDog/ebpf-manager"
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/features"
+	"github.com/cilium/ebpf/perf"
+	"github.com/cilium/ebpf/ringbuf"
+
+	ebpfTelemetry "github.com/DataDog/datadog-agent/pkg/ebpf/telemetry"
+	ddsync "github.com/DataDog/datadog-agent/pkg/util/sync"
+)
+
+var perfPool = ddsync.NewDefaultTypedPool[perf.Record]()
+var ringbufPool = ddsync.NewDefaultTypedPool[ringbuf.Record]()
+
+// Flushable is an interface for objects that support flushing
+type Flushable interface {
+	Flush()
+}
+
+// EventHandler abstracts consuming data from a perf buffer or ring buffer (depending on availability and options).
+// It handles upgrading maps from a ring buffer is desired, and unmarshalling into the desired data type.
+type EventHandler struct {
+	f    Flushable
+	opts EventHandlerOptions
+}
+
+// EventHandlerOptions are the options controlling the EventHandler.
+// MapName and Handler are required options.
+type EventHandlerOptions struct {
+	// MapName specifies the name of the map. This field is required.
+	MapName string
+	// Handler is the callback for data received from the perf/ring buffer. This field is required.
+	Handler func([]byte)
+
+	// TelemetryEnabled specifies whether to collect usage telemetry from the perf/ring buffer.
+	TelemetryEnabled bool
+	// UseRingBuffer specifies whether to use a ring buffer
+	UseRingBuffer bool
+	// UpgradePerfBuffer specifies if you wish to upgrade a perf buffer to a ring buffer.
+	// This only takes effect if UseRingBuffer is true.
+	UpgradePerfBuffer bool
+
+	PerfOptions    PerfBufferOptions
+	RingBufOptions RingBufferOptions
```

Comment:
> are those 2 mutually exclusive when creating the EventHandlerOptions instance?

They are mutually exclusive. Let me think about a way to encode that. It may be difficult with Go's type system.

Reply:
They are actually not mutually exclusive! It is possible to optionally upgrade to ring buffers if available.

Reply:
I reworked the API, let me know what you think.